tutor/src/components/annotations/annotation.jsx

```diff
@@ -0,0 +1,373 @@
+
+  }
+
+  componentDidMount() {
+    this.handleSelectionChange = debounce(() => {
+      if (!this.activeHighlight) {
+        this.setWidgetStyle();
+      }
+    }, 80);
+    this.props.windowImpl.document.addEventListener('selectionchange', this.handleSelectionChange)
+    this.restoreAnnotations();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.needToLoadAnnotations = nextProps.documentId !== this.props.documentId;
+  }
+
+  componentDidUpdate() {
+    this.restoreAnnotations();
+  }
+
+  componentWillUnmount() {
+    this.props.windowImpl.document.removeEventListener('selectionchange', this.handleSelectionChange)
+  }
+
+  highlightEntry(entry) {
+    const selection = serializeSelection.restore(entry.selection, this.articleElement);
+
+    if (! ('rect' in entry)) {
+      entry.rect = getSelectionRect(selection);
+    }
+    highlighter.doHighlight();
+  }
+
+  unhighlightEntry(entry) {
+    // Highlight the entry, then unhighlight the highlights with the latest timestamp
+    this.highlightEntry(entry);
+    const highlights = highlighter.getHighlights({grouped: true});
+    const lastHighlight = highlights.reduce((a, b) => a.timestamp > b.timestamp ? a : b);
+
+    for (const el of lastHighlight.chunks) {
+      highlighter.removeHighlights(el);
+    }
+  }
+
+  findOverlap() {
+    const selection = serializeSelection.save(this.articleElement);
+    const start = selection.start;
+    const end = selection.end;
+
+    return this.preload.find((entry) => {
+      const sel = entry.selection;
+
+      return (sel.start >= start && sel.start <= end) ||
+        (sel.end >= start && sel.end <= end)
+    });
+  }
+
+  @action.bound
+  setWidgetStyle() {
+    const selection = this.props.windowImpl.getSelection();
+
+    // If it's a cursor placement with no highlighted text, check
+    // for whether it's in an existing highlight
+    if (selection.isCollapsed) {
+      this.widgetStyle = null;
+      this.savedSelection = null;
+      const start = serializeSelection.save(this.articleElement).start;
+      this.activeHighlight = this.preload.find((entry) => {
+        const sel = entry.selection;
+
+        return sel.start <= start && sel.end >= start;
+      })
+
+      // Set selection, which will cause the widget to render
+      if (this.activeHighlight) {
+        serializeSelection.restore(this.activeHighlight.selection, this.articleElement);
+      }
+    } else if (this.findOverlap()){
+      console.warn("NO OVERLAPPING SELECTIONS!");
+      this.widgetStyle = null;
+    } else {
+      this.savedSelection = serializeSelection.save(this.articleElement);
+      const rect = getSelectionRect(selection);
+      const pwRect = this.parentRect;
+
+      this.widgetStyle = {
+        top: `${rect.bottom - pwRect.top}px`,
+        left: `${rect.left - pwRect.left}px`
+      };
+    }
+  }
+
+  restoreAnnotations() {
+    if (this.needToLoadAnnotations) {
+      hypothesisStore.fetch(this.props.documentId).then(action(
```

Comment:
Does tutor's version of JS allow us to use `async` rather than promises?

Reply:
I **think** it does, we're using the `stage-1` preset.   We haven't used them elsewhere yet so up to @RoyEJohnson if he wants to be the first 😁